Rebase density approximations

gassmoeller opened the pull request:
This is a rebase of #954 to check for test failures and to base on a recent master. It needs more work, but it is our suggestion on how BA, TALA, ALA and possibly further approximations can be implemented. I will add more information when the PR is in a reviewable state.

bangerth commented:
Ouch, 99.25% of tests failed :-)

I think this deserves the @egpuckett "Puckett bucket" award for November ;-)

gassmoeller commented:
I will defend myself by blaming @tjhei :smile:. He wanted to continue working, while Juliane and I were travelling so we uploaded the unfinished branch. The interesting question (which I did not investigate) is why there were tests that passed with an invalid default input parameter that is currently not set in any parameter file. Maybe those were the newly added tests, or the ones that were expected to fail anyway.

tjhei commented:
> Maybe those were the newly added tests, or the ones that were expected to fail anyway.

yes, we have a few tests that are expected to fail.

tjhei commented:
Rene, we currently have ``parameters.formulation_compressibility`` which is set by "mass density formulation" in the prm. I think the names should be more similar. I also think both are confusing (at least to me). What do you think about using "mass conservation" and ``parameters.formulation_mass_conservation``?

tjhei commented:
TanGurnis and King benchmarks are updated and working. I also added test coverage for them. I think we have most of what is needed to merge this PR. Thoughts?

bangerth commented:
Tell me if this is ready for review!

gassmoeller commented:
I thought about the naming for a while, but I really do not have a perfect solution. 
In my mind naming it something with `compressibility formulation` (both in code and parameter file) makes most sense, because it really affects the term that shows up in compressible models (I do not count the -1/3 term in the momentum equation, because that one technically is also there for the incompressible models, it just happens to be zero). 
All names that include `mass_density...` do not make it clear that it means the `mass conservation equation density gradient approximation` and to use the full term is just not practical. @bangerth suggested to maybe hide it in further namespaces/subsections, e.g. `Formulation::ContinuityEquation::DensityApproximation`. This would be clearer, I am just afraid of the practical consequences for the code (think of `if (parameters.continuity_density_approximation == Parameters<dim>::Formulation::ContinuityEquation::DensityApproximation::implicit_reference_profile) ...`
Maybe there is not a best way, and we just need to decide for something.

bangerth commented:
Afraid of long names? NOT ME!  ;-)

tjhei commented:
> "compressibility formulation"

At least to me it is not obvious what that changes especially in comparison with ``Material::is_compressible``. Because of that I very much prefer a name that reflects which of the equations is modified (so "mass conservation" in this case).

> suggested to maybe hide it in further namespaces/subsections, e.g. Formulation::ContinuityEquation::DensityApproximation

But you are not only changing how the density is approximated, but also the density gradient, etc..

What about
```
subsection formulation
set mass conservation = incompressible | reference density profile | isothemal compression | ...
set temperature equation = real density | reference density profile
end
```
being mapped to ``.formulation_mass_conservation`` and ``.formulation_temperature``?

egpuckett commented:
How about Equation(s) of State (EOS)? 

> On Nov 29, 2016, at 12:33 PM, Timo Heister <notifications@github.com> wrote:
> 
> "compressibility formulation"
> 
> At least to me it is not obvious what that changes especially in comparison with Material::is_compressible. Because of that I very much prefer a name that reflects which of the equations is modified (so "mass conservation" in this case).
> 
> suggested to maybe hide it in further namespaces/subsections, e.g. Formulation::ContinuityEquation::DensityApproximation
> 
> But you are not only changing how the density is approximated, but also the density gradient, etc..
> 
> What about
> 
> subsection formulation
> set mass conservation = incompressible | reference density profile | isothemal compression | ...
> set temperature equation = real density | reference density profile
> end
> being mapped to .formulation_mass_conservation and .formulation_temperature?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

gassmoeller commented:
@tjhei I agree, and convinced myself that your suggestion is the best available option :smile:. I will go ahead and change the code  and commited parameter files accordingly.

@egpuckett Equation of State might be an alternative, but the specific equation of state that is used is determined by the material model, or the adiabatic conditions, but not by our choice which terms to include in the equation (of course we try to enforce that they match, it makes no sense to use an incompressible equation of state (=material model) in a set of  compressible equations).

egpuckett commented:
@gassmoeller I think I understand your reasons for thinking EOS is a not good choice, and perhaps it's not. I suggested EOS, since this is how my colleagues and I handled this same question in the past for finite difference codes for modeling hyperbolic conservation laws, as in gas dynamics.  One can think of the EOS as the equation (or equations) that close the set of equations that are explicitly `solved' in the code. For example, for BA, we have **u** = (u,v,w), ρ, P and T with a additional thermodynamic and material constants such as c_p, ղ, etc. that may vary with space and / or time and other quantities such as temperature. The incompressible BA model solves for **u** = (u,v,w), P, and T, GIVEN ρ.  However, we don't have a PDE or ODE for ρ, (i.e., an equation for updating ρ in time) so we use the Boussinesq approximation 

> ρ(**x** , t) = ρ_0 ( 1 - α_v (T − T_0 ))

From a certain point of view, this is an 'equation of state' for the most basic 'mantle convection' equations one can approximate in  ASPECT: incompressible Stokes and an advection-diffusion equation for the temperature, plus an `equation-of-state', namely the Boussinesq approximation.

Perhaps a more straightforward (or universally accepted) example is gas dynamics or, more generally, a system of hyperbolic conservation laws. In 1D one has three conservation laws: mass, linear momentum, and energy, but four independent variables, ρ, u, e and p, where e is the internal energy and p is the (thermodynamic) pressure. The code will have some method for advancing the three conserved quantities ρ (mass density), m = ρ u (momentum), E = ρ e + ρ u*u / 2 (total energy) forward in time, but one has to determine the pressure p from the other variables in order to do this; i.e., in order to `close the system of equations'. **NOTE:** Just as in the mantle convection equations, there is no time dependent equation for the pressure; i.e., dp / dt does not appear anywhere in the equations.   Instead, one finds the pressure by  using an equation of state.  In gas dynamics, for a gamma law gas EOS, one uses

> p = ρ (γ -1) e

where γ is the ratio of specific heats γ = c_p / c_v, which is fixed for a given gas under reasonable conditions.

However, there are a very large number of other ways to define the  p in terms of the other variables, including table look ups, mie-Gruneisen EOS, and other approaches that may also include adding additional thermodynamic or material variables and / or constants, depending on the problem.  One can even model 'solids' with the appropriate EOS, i.e., equation or equations that give the pressure p in terms of the other variables that are obtained from the solution of the (hyperbolic) conservation laws.

One caveat: Compressible and incompressible formulations differ in a fundamental way, since in an incompressible formulation the density ρ does't explicitly depend on time in a PDE, and the pressure,  is not the thermodynamic pressure, but just the dynamic pressure, which comes along for the ride - so to  speak.

Anyway, I don't have a strong feeling about the naming convention you chosen for these models, but I do want to explain my reasoning for using something like EOS as a way of explaining to the user what that part of the code / parameter file 'is / does' and see if I could get some feedback.  @jdannberg @tjhei? others?

BTW, I thought about 'constitutive model' or something like that, but my understanding is that this implies so-called 'strength' effects, the sort of thing John Naliboff, Cedric, and others are working on.

gassmoeller commented:
I agree with everything you said :smile:. I guess my point was mostly that EOS is a good name for the whole concept of how to relate the material properties in the equations and to close the system, but we were looking for the name for the formulation of one of the equations (mass conservation in this case).
For now I renamed everything in the way Timo suggested, cleaned the code a bit, and added checks for the consistency of input parameters. 
@bangerth The code should be ready for a review.

Things still to do:
- update the manual

tjhei commented on source/simulator/core.cc:
```diff
@@ -642,6 +642,9 @@ namespace aspect
         prm.print_parameters(prm_out, ParameterHandler::LaTeX);
       }
 
+    // check that the setup of equations, material models, and heating termes is consistent
```
"terms"

tjhei commented on source/simulator/helper_functions.cc:
```diff
@@ -1320,6 +1323,104 @@ namespace aspect
+      {
+        const std::vector<std::string> &heating_models = heating_model_manager.get_active_heating_model_names();
+
+        AssertThrow(std::find(heating_models.begin(), heating_models.end(), "adiabatic heating") != heating_models.end(),
```
Would it make sense to introduce ``heating_model_manager.adiabatic_heating_enabled()`` and similarly for shear heating? It makes the assert easier to read and makes it more robust (one location, not based on a string comparison, allows future alternative plugins, etc.).

tjhei commented on source/simulator/helper_functions.cc:
```diff
@@ -1320,6 +1323,104 @@ namespace aspect
+
+  template <int dim>
+  void
+  Simulator<dim>::check_consistency_of_formulation() const
```
very nice!

tjhei commented on source/simulator/assemblers/stokes.cc:
```diff
@@ -335,8 +335,11 @@ namespace aspect
+    Assert((parameters.formulation_mass_conservation ==
+            Parameters<dim>::FormulationMassConservation::isothermal_compression) ||
+           ((parameters.formulation_mass_conservation ==
+             Parameters<dim>::FormulationMassConservation::ask_material_model) &&
```
I think the code would be much easier if we replace ask_material_model at initialization time (or if this is too early in your validation routine).

tjhei commented on tests/material_model_dependencies_tangurnis.prm:
```diff
@@ -88,9 +88,16 @@ subsection Mesh refinement
 subsection Model settings
-  set Include adiabatic heating               = false
-  set Include shear heating                   = false # default: true
+  set Include adiabatic heating               = true
```
this is the deprecated way to set the plugins, right? Can you change that?

tjhei commented on source/simulator/helper_functions.cc:
```diff
@@ -1320,6 +1323,104 @@ namespace aspect
+        for (typename std::list<std_cxx11::shared_ptr<HeatingModel::Interface<dim> > >::const_iterator plugin =
+               heating_plugins.begin(); plugin != heating_plugins.end(); ++plugin)
+          {
+            const HeatingModel::AdiabaticHeating<dim> *
```
This would also make for a nice helper function in ``heating_model_manager``, for example ``HeatingModel::AdiabaticHeating<dim> * = heating_model_manager.get_adiabatic_heating_plugin()``.

gassmoeller commented:
I addressed the comments and rebased to a current master.
I would like to separate the replacement of the deprecated parameter file option to a different PR (a quick search found more than 500 places where we still use this parameter). I need a script for that :smile:.

tjhei commented:
> I would like to separate the replacement of the deprecated parameter file option to a different PR

Agreed. I will take a look at my tests...

tjhei commented:
We are green now, yay!

gassmoeller commented:
I will do a rebase tomorrow, which will break the tests again :smile:. But that should be simple to solve.

bangerth commented:
There is a merge conflict again, for reasons not clear to me. Can you take another look?

bangerth commented on benchmark/king2dcompressible/ala/README:
```diff
@@ -0,0 +1,18 @@
+different values of the Dissipation number Di and the Rayleight 
+number Ra. The complete input file can be found in 'ala.prm'. 
+All other input files only contain the modifications necessary 
+ro run cases with a different approximation or values of Di and Ra, 
```
ro -> to

bangerth commented on benchmark/king2dcompressible/ala/README:
```diff
@@ -0,0 +1,18 @@
+The bash script 'get_statistics' can copy the relevant statistics
+of one or several model runs to the screen output and a separate
+file. It takes one or several paths to a number of statistics 
+files as input. 
```
That description is too vague for me to really understand what the script is doing.

bangerth commented on benchmark/king2dcompressible/ala/ala.prm:
```diff
@@ -0,0 +1,286 @@
+# CU   10^4 0.25 4.4100 40.0000 58.1000   38.8000  0.5148 0.8494 0.8501
+# wealt   "    "    4.44   39.9                       0.51493 8.46633256e-01 8.46071386e-01 
+# we   "    "    4.43957   39.887840                       0.514951021 8.46633256e-01 8.46071386e-01 
+# okay
```
???

bangerth commented on benchmark/king2dcompressible/ala/ala.prm:
```diff
@@ -0,0 +1,286 @@
+
+# CU   10^5 0.25 9.2100 178.2000 252.4000 184.3000 0.5319 2.0503 2.0540
+# WE   "    "    9.50829 178.07837                  0.532474458   2.05198146e+00  -2.05114850e+00  
+# okay
```
???

bangerth commented on benchmark/king2dcompressible/ala/ala.prm:
```diff
@@ -0,0 +1,286 @@
+  subsection Function
+#    set Function constants  = z1=0.102367, z2=0.897633, pi=3.14159265359
+#    set Function expression = if(z<z1,0.5+0.5*(z1-z)/z1,if(z>z2,0.5*(1-z)/(1-z2),0.5)) + 0.1 * cos(x*pi) * sin(z*pi)
+#    set Variable names      = x,z
```
as in a previous .prm file, what actually happens if you comment all of this out, but still select `function`?

bangerth replied:
Oh, I see. Awkward. In that case, please just merge the sections and remove the commented out parts.

bangerth commented on benchmark/king2dcompressible/ala/ala.prm:
```diff
@@ -0,0 +1,286 @@
+  # adiabatic heating: alpha T (u . \nabla p)
+  # 
+
+  set List of model names = adiabatic heating, shear heating#, constant heating
```
remove commented out things, or be more verbose what they're supposed to mean

bangerth commented on benchmark/king2dcompressible/ala/ala.prm:
```diff
@@ -0,0 +1,286 @@
+  subsection Adiabatic heating
+    # A flag indicating whether the adiabatic heating should be simplified
+    # from $\alpha T (\mathbf u \cdot \nabla p)$ 
+    # to $ \alpha \rho T (\mathbf u \cdot \mathbf g) $. alpha=thermal exp
```
same in this section

bangerth commented on benchmark/king2dcompressible/ala/ala.prm:
```diff
@@ -0,0 +1,286 @@
+subsection Formulation
+  set Formulation = ALA # equivalent to 'custom' with the settings below
+  #set Mass conservation    = reference density profile
+  #set Temperature equation = reference density profile
```
Do either, but not both: choose ALA and remove the two other lines, or select "custom" and leave them in. Otherwise, I think it is just too confusing; if you want to discuss the equivalence, the manual is a better place than a half-line comment in a .prm file.

bangerth commented on benchmark/king2dcompressible/ala/ala.prm:
```diff
@@ -0,0 +1,286 @@
+
+    # The value of the constant viscosity. Units: $kg/m/s$.
+    set Ra                     = 1e4
+    set Di = 1.0 # BA: Di=0.0
```
same comment as before

bangerth commented on benchmark/king2dcompressible/ala/ala.prm:
```diff
@@ -0,0 +1,286 @@
+    set Di = 1.0 # BA: Di=0.0
+    set b=0
+    set c=0
+    set gamma = 1.0
```
maybe align the `=` signs in this section for better readability?

bangerth commented on benchmark/king2dcompressible/ala/ala.prm:
```diff
@@ -0,0 +1,286 @@
+end
+
+
+subsection Material model
```
move section up to the top?

bangerth commented on benchmark/king2dcompressible/ala/ala.prm:
```diff
@@ -0,0 +1,286 @@
+end
+
+
+# The final part is to specify what ASPECT should do with the
```
This an several of the other files repeat a lot of commentary from existing .prm files. I don't know whether that's necessary -- I'll leave this up to you, just wanted to point out the duplication.

bangerth commented on benchmark/king2dcompressible/ala/ala_Di025_Ra4.prm:
```diff
@@ -0,0 +1,23 @@
+end
+
+subsection Material model
+  subsection Tan Gurnis model
```
Does it really use the Tan-Gurnis model, or was that just a convenient starting point?

bangerth commented on benchmark/king2dcompressible/ala/code.cc:
```diff
@@ -0,0 +1,284 @@
+  /**
+   * This benchmark is from the article
+   * @code
+   * @endcode
```
Make same changes as in the other `code.cc` files. In fact, you'd think that there ought to be ways to reduce code duplication. I haven't looked through the files in detail, but it seems like there is a significant amount of common code, right?

bangerth commented on benchmark/king2dcompressible/ala/get_statistics:
```diff
@@ -0,0 +1,14 @@
+#!/bin/bash
+
```
This script could use a comment about how it's supposed to be used (or where the description of how it's supposed to be used is located).

bangerth commented on benchmark/king2dcompressible/ala/results.txt:
```diff
@@ -0,0 +1,45 @@
+#Di  Ra  Nu             Vrms           <T>      phi            W
+
+0.25 1e4 
+0.25 1e4 
```
Are the missing columns still coming?

bangerth commented on benchmark/king2dcompressible/eba/code.cc:
```diff
@@ -0,0 +1,447 @@
+  /**
+   * This benchmark is from the article
+   * @code
+   * @endcode
```
same comments as for the other files.

bangerth commented on benchmark/king2dcompressible/eba/eba1.prm:
```diff
@@ -0,0 +1,253 @@
+# reference values: 
+# Ra=1e4, Di 0.25, , Nu 4.0900, RMS 38.4000
+# -> ref4: 4.323 | 39.4
+#    ref5: 4.238 | 39.21
```
same comments as for the other .prm files

bangerth commented on benchmark/tangurnis/Makefile:
```diff
@@ -6,10 +6,11 @@ tala_c: tala_c/vel_8.csv tala_c/vel_16.csv tala_c/vel_32.csv tala_c/vel_64.csv t
 
 
-%.csv: ../../aspect ba/tan.prm tala/tan.prm tala_c/tan.prm
+%.csv: Makefile code/aspect ba/tan.prm tala/tan.prm tala_c/tan.prm
```
here and below, does it make sense to hardcode the path to `aspect`? If you want to keep it variable, maybe define the path to the binary somewhere at the top as a makefile variable.

jdannberg replied:
@tjhei maybe you can comment on this?
I think I've addressed all other comments about the king and tan gurnis benchmarks [here](https://github.com/jdannberg/aspect/commits/density_approximations).
In particular, I cleaned up the code and also updated the comments in the input files.

bangerth commented on benchmark/tangurnis/code/tangurnis.cc:
```diff
@@ -303,20 +305,24 @@ namespace aspect
+
+      // this is no longer used because we use the new adiabatic mass formulation
+      // based on AdiabaticConditions
       return (d==0) ? 1.0 : (Di/gamma / d);
```
If no longer used, maybe just return `numbers::signaling_Nan<double>()`?

bangerth commented on benchmark/tangurnis/code/tangurnis.cc:
```diff
@@ -581,43 +537,92 @@ namespace aspect
+        fe_values[this->introspection().extractors.temperature].get_function_values (this->get_solution(), in.temperature);
+        fe_values[this->introspection().extractors.pressure].get_function_values (this->get_solution(), in.pressure);
+        fe_values[this->introspection().extractors.velocities].get_function_values (this->get_solution(), in.velocity);
+        fe_values[this->introspection().extractors.pressure].get_function_gradients (this->get_solution(), in.pressure_gradient);
```
Do you recall where you need the pressure gradient? As discussed on some other occasion, I think that any formulation that actually needs the pressure gradient is automatically a bit suspect.

jdannberg replied:
We might not even need it in this specific case, but the pressure gradient is needed for example in the version of the adiabatic heating that is used by default if the user does not choose to use the simplified adiabatic heating. Hence, everywhere where we compute heating model outputs we normally also fill the pressure gradients. 
We can think about changing this in the future, but it is used in many other places in the code and probably something for a different pull request.

bangerth commented on benchmark/tangurnis/matlab.m:
```diff
@@ -28,7 +34,7 @@
         k = vv(1,3)*pi; %1*pi
         a=vv(1,4); % 0 or 2
-        fprintf('computing Di=%i, gamma = %i\n', Di, gamm)
+        fprintf('computing Di=%i, gamma = %i, a= %i\n', Di, gamm, a)
```
Who wrote this originally that they couldn't spare the extra letter to call it `gammA`? ;-)

tjhei replied:
I did. ``gamma`` is a function in MATLAB so you can not use it as a variable.

bangerth replied:
Ha, that's funny. I did not know :-)

bangerth commented on benchmark/tangurnis/notes.txt:
```diff
@@ -1 +1,14 @@
+analytical solution (Tan, Gurnis '07)  5.306916e-03   5.251999e-03
+
+numerical solution (refinement 4)      5.30690178e-03 5.25196977e-03
+numerical solution (refinement 5)      5.30691533e-03 5.25199721e-03
```
Oh, excellent match!

bangerth replied:
But also what an interesting testcase where you get the right answer to five digits already on the coarsest mesh. Maybe not the best chosen output quantity then...

jdannberg replied:
This is just to test that the terms are implemented correctly. We compare the heating terms in the first time step (with time step length 0) to the analytical solution, so we don't really test the temperature solver, we just check if we have the correct RHS terms. But this way we can check that our approximations work exactly in the way other people have implemented them, and it's very easy to find where the problem is in case something changes in the future.

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1265,9 +1265,9 @@ \subsection{Numerical methods}
   location of plumes that require high resolution change and move with time,
   meshes also need to be adapted every few time steps.
-\item \textit{Accurate discretizations:} The Boussinesq problem upon which
+\item \textit{Accurate discretizations:} The equations upon which
```
good catch -- we need to purge this term from the manual

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1315,14 +1316,233 @@ \subsection{Numerical methods}
+approximation (ALA), truncated anelastic liquid approximation (TALA), and the
+Boussinesq approximation (BA). These can all be derived from the basic
+equations~\eqref{eq:stokes-1}--\eqref{eq:temperature} via various approximation, though
+there are a number of pitfalls if one wanted to use them in \aspect{}. We will
```
wanted -> wants?

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1315,14 +1316,233 @@ \subsection{Numerical methods}
+pressure fields from this reference state. In particular, the reference pressure
+is given by solving the hydrostatic equation,
+\begin{align*}
+  \nabla \bar p = \bar\rho \mathbf g,
```
Is this the case for all three approximations?

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1315,14 +1316,233 @@ \subsection{Numerical methods}
+  \nabla \bar p = \bar\rho \mathbf g,
+\end{align*}
+where $\bar\rho=\rho(\bar p,\bar T)$ is a \textit{reference density} that
+depends on depth. $\bar T$ is chosen as an adiabatic profile accounting for the
```
Does it depend on depth even for the incompressible models? Or only for some of the 3 approximations, in which case I would enumerate them here.

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1315,14 +1316,233 @@ \subsection{Numerical methods}
+fact that the temperature increases as the pressure increases, but also taking
+\marginpar{Scott: I don't see this discussed in \cite{KLKLZTTK10}, but one
+needs to make $\bar T$ also take into account diffusion so that it disappears
+from the temperature equation in the ALA.} 
```
Did you mean to leave this comment in, or was that a left-over from earlier discussions?

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1315,14 +1316,233 @@ \subsection{Numerical methods}
+\marginpar{Scott: I don't see this discussed in \cite{KLKLZTTK10}, but one
+needs to make $\bar T$ also take into account diffusion so that it disappears
+from the temperature equation in the ALA.} 
+into account thermal diffusion effects. It also depends on depth.
```
Do you mean to say that it *explicitly* depends on the depth? If yes, I would say "explicitly", otherwise, I would say "Consequently, through the dependence on the pressure, it varies with depth."

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1315,14 +1316,233 @@ \subsection{Numerical methods}
+  \textrm{in $\Omega$}.
+\end{multline}
+
+Starting from these equations, the approximations discussed in the next few
```
You've already used "starting from" in the previous paragraph. Maybe "Based on these equations, ..."

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1315,14 +1316,233 @@ \subsection{Numerical methods}
+almost always given in terms of non-dimensionalalized quantities. We will for
+now stick with the dimensional form because it expresses in a clearer way the
+approximations that are made. The non-dimensionalization can then be done on
+each of the forms below separately.
```
Yes!

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1315,14 +1316,233 @@ \subsection{Numerical methods}
+\end{align*}
+Here, $\frac{\partial \rho(\bar p,\bar T)}{\partial T}$ is related to the
+thermal expansion coefficient, and $\frac{\partial \rho(\bar p,\bar T)}{\partial p}$ to the
+compressibility.
```
Is it worthwhile stating how exactly they are related?

bangerth replied:
I still think so :-)

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1315,14 +1316,233 @@ \subsection{Numerical methods}
+thermal expansion coefficient, and $\frac{\partial \rho(\bar p,\bar T)}{\partial p}$ to the
+compressibility.
+
+The second assumotion is that the variation of the density
```
-> assumption

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1315,14 +1316,233 @@ \subsection{Numerical methods}
+  & \qquad
+  & \textrm{in $\Omega$}.
+\end{align}
+\marginpar{Timo: need to figure out how to solve \eqref{eq:stokes-ALA-2}, need $\nabla \rho$! }
```
what do you do right now in the code?

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1315,14 +1316,233 @@ \subsection{Numerical methods}
+\marginpar{Timo: need to figure out how to solve \eqref{eq:stokes-ALA-2}, need $\nabla \rho$! }
+
+For the temperature equation we arrive at the following:
+\marginpar{Scott: This does not match (18) in \cite{KLKLZTTK10}}
```
has this been resolved?

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1315,14 +1316,233 @@ \subsection{Numerical methods}
+  \quad
+  \textrm{in $\Omega$}.
+\end{multline}
+Timo: this is not quite correct, you need reference density and simplified adiabatic heating to get:
```
highlight this somehow so we don't forget the comment

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1315,14 +1316,233 @@ \subsection{Numerical methods}
+\subsubsection{The truncated anelastic liquid approximation (TALA)}
+
+The \textit{Truncated Anelastic Liquid Approximation (TALA)} further simplifies
+the ALA by assuming that the variation of the density due to pressure variations
```
too many "variations" :-) Maybe say "that the variation of the density due to $p'$ is small"?

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1315,14 +1316,233 @@ \subsection{Numerical methods}
+example, continue to be depth dependent because the hydrostatic pressure grows
+with depth. It simply means that the deviations from the reference pressure are
+assumed to be so small that they do not matter in describing the density.
+Because the pressure variation is induced by the flow field (the static
```
-> pressure variation $p'$

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1315,14 +1316,233 @@ \subsection{Numerical methods}
+  \nabla \cdot (\bar\rho \mathbf u) &= 0
+  & \qquad
+  & \textrm{in $\Omega$}.
+\end{align}
```
does the TALA have any impact on the temperature equation? If not, maybe say so.

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1315,14 +1316,233 @@ \subsection{Numerical methods}
+
+These equations are identical in form for the \textit{Boussinesq approximation
+(BA)} which only differs from the EBA in neglecting compressibility effects in
+the temperature equation (shear heating and simplified adiabatic heating).
```
I think it would be useful to state the temperature equation here for both the EBA and the BA.

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1315,14 +1316,233 @@ \subsection{Numerical methods}
+
 
 \subsubsection{The Boussinesq approximation: Incompressibility}
 \label{sec:boussinesq}
```
Do we then still need this section?

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1578,6 +1798,145 @@ \subsubsection{Compressible models}
+\newpage
+\subsection{Approximations}
+
+There are several common simplifications of the above mentioned equations. They usually rely on the computation of a depth-dependent reference profile that represents a typical change of material parameters and solution variables with depth. The approximations are justified by the fact that the actual conditions do not deviate much from this reference profile. The terms influenced by these approx ...
```
what about this section?

tjhei commented on benchmark/blankenbach/case2/ReadMe.txt:
```diff
@@ -0,0 +1,24 @@
+* case2a:
```
The Blankenbach stuff is still somewhat incomplete. Maybe we should do this in a separate PR.

bangerth replied:
Your choice. But I think the comments I had should be addressed whether or not this all is included in the current patch or not.

bangerth commented on include/aspect/assembly.h:
```diff
@@ -119,6 +116,9 @@ namespace aspect
           MaterialModel::MaterialModelOutputs<dim> face_material_model_outputs;
+
+          std::vector<double> mass_densities;
+          std::vector<double> adiabatic_density_gradients;
```
Can you add a comment what these correspond to? I'm also slightly confused by the fact that the second variable is a *gradient* but has data type `double`. A comment would probably be helpful.

bangerth commented on include/aspect/heating_model/adiabatic_heating.h:
```diff
@@ -54,6 +54,13 @@ namespace aspect
+         * neglecting dynamic pressure effects.
+         */
+        bool
+        use_simplified_adiabatic_heating() const;
```
I don't really know what this means in formulas. It would be great to include the two forms of the term in the documentation, depending on whether or not  the simplified form is used.

bangerth commented on include/aspect/heating_model/interface.h:
```diff
@@ -298,6 +323,30 @@ namespace aspect
+        if (HeatingModelType *x = dynamic_cast<HeatingModelType *> ( (*p).get()) )
+          return x;
+      return NULL;
+    }
```
Nice. If you're enterprising, the changes to this file could also easily go into a separate patch.

bangerth commented on include/aspect/parameters.h:
```diff
@@ -202,6 +202,109 @@ namespace aspect
+     * @name Formulation settings
+     * @{
+     */
+    struct CombinedFormulation
```
I think this might actually be easier to do with a `namespace`. This way you don't have to mark the function as `static` (though you then either need to put it into the .cc file -- my preference -- or make it `inline`).

bangerth replied:
Some more documentation would be nice to have, both for the struct/namespace and for its members, to explain what they are used for and what they concretely do.

bangerth commented on include/aspect/parameters.h:
```diff
@@ -202,6 +202,109 @@ namespace aspect
+      }
+    };
+
+    struct FormulationMassConservation
```
same here. some documentation would be good.

bangerth commented on include/aspect/parameters.h:
```diff
@@ -202,6 +202,109 @@ namespace aspect
+      }
+    };
+
+    struct FormulationTemperatureEquation
```
and here

bangerth commented on include/aspect/parameters.h:
```diff
@@ -202,6 +202,109 @@ namespace aspect
+     * original formulation is termed 'isothermal compression'. 'Custom' allows
+     * to set the approximations individually.
+     */
+    typename CombinedFormulation::Kind combined_formulation;
```
No need for the `typename`. Same below.

bangerth replied:
The comment has a kind of weird start: "A consistent way to formulate the equations to solve." This doesn't reference what the variable actually does. Maybe "This variable selects one of the consistent ways to formulate..."

bangerth commented on include/aspect/parameters.h:
```diff
@@ -202,6 +202,109 @@ namespace aspect
+      {
+        BA,
+        ALA,
+        isothermal_compression,
```
`isothermal_compressible`?

bangerth commented on include/aspect/parameters.h:
```diff
@@ -202,6 +202,109 @@ namespace aspect
+    {
+      enum Kind
+      {
+        isothermal_compression,
```
same here

bangerth commented on include/aspect/parameters.h:
```diff
@@ -202,6 +202,109 @@ namespace aspect
+      };
+      static Kind parse(const std::string &input)
+      {
+        if (input == "isothermal compression")
```
is necessary, adjust here as well

bangerth commented on include/aspect/parameters.h:
```diff
@@ -202,6 +202,109 @@ namespace aspect
+          return FormulationTemperatureEquation::reference_density_profile;
+        else
+          AssertThrow(false, ExcNotImplemented());
+        return FormulationTemperatureEquation::real_density;
```
In this and the other functions, what you really want to do is return an invalid value at this point. I would just write
```
  return FormulationTemperatureEquation::Kind();
```
to make this clear.

bangerth commented on include/aspect/simulator.h:
```diff
@@ -1140,6 +1139,18 @@ namespace aspect
+       * selection is detected it throws an exception. It for example assures that
+       * correct heating terms are selected, and the material model supports
+       * the selection of the mass conservation formulation (e.g. incompressible)).
+       * If 'parameters.formulation' is set to 'custom'
```
There are now multiple of these `.formulation` variables.

bangerth commented on include/aspect/simulator_access.h:
```diff
@@ -64,6 +64,10 @@ namespace aspect
+  namespace AdiabaticConditions
+  {
+    template <int dim> class Interface;
+  }
```
How did we not need this before? You're not making any other changes to this file...

gassmoeller commented:
I will go ahead and address the comments in the code, and start restructuring the manual. @tjhei and @jdannberg maybe you can take a look at the benchmarks and address those comments (and decide if the Blankenbach cases should go in a separate PR to be merged after this one). I will wait with the rebase until we addressed the changes, otherwise it is not clear which ones are sufficiently addressed already.

bangerth commented on source/simulator/entropy_viscosity.cc:
```diff
@@ -0,0 +1,439 @@
+        if (parameters.formulation_temperature_equation ==
+            Parameters<dim>::FormulationTemperatureEquation::reference_density_profile)
+          for (unsigned int q=0; q<n_q_points; ++q)
+            scratch.material_model_outputs.densities[q] = adiabatic_conditions->density(scratch.material_model_inputs.position[q]);
```
I think you should probably do this loop before the call to `MaterialAveraging::average` above. The same may be true in other assemblers but I didn't pay attention there -- please double check that the averaging always happens *after* the correct densities have been computed.

bangerth commented on include/aspect/parameters.h:
```diff
@@ -202,6 +202,109 @@ namespace aspect
+        reference_density_profile,
+        implicit_reference_density_profile,
+        incompressible,
+        ask_material_model
```
I think that from the use later on, you use `ask_material_model` as a "yet-to-be-determined" value, right? Maybe call it that way.

gassmoeller replied:
Would you be ok with `fallback_to_material_model`?  I would like to make clear where the value comes from, and preferentially also that it is more of a deprecated option that is not our preferred way.

bangerth commented on source/simulator/assembly.cc:
```diff
@@ -2358,27 +637,79 @@ namespace aspect
+                 Parameters<dim>::FormulationMassConservation::incompressible
+                 || (parameters.formulation_mass_conservation ==
+                     Parameters<dim>::FormulationMassConservation::ask_material_model
+                     && !material_model->is_compressible()))
```
Is `ask_material_model` allowed here because the formulation does not matter if the material model says that it's not compressible anyway?

I would prefer if we simply asserted that we never encounter `ask_material_model` at all...

gassmoeller replied:
No this was a leftover. `ask_material_model` is already replaced by the appropriate values in `check_consistency_of_formulation`. In the long term we would like to remove `is_compressible` from any decision on the assembled terms and only use it for consistency checking (if at all).

bangerth commented on source/simulator/helper_functions.cc:
```diff
@@ -1320,6 +1323,106 @@ namespace aspect
+        AssertThrow(material_model->is_compressible() == false,
+                    ExcMessage("ASPECT detected an inconsistency in the provided input file. "
+                               "The mass conservation equation was selected to be incompressible, "
+                               "but the provided material model reports to be compressible. "
```
-> reports that it is compresible

bangerth commented on source/simulator/helper_functions.cc:
```diff
@@ -1320,6 +1323,106 @@ namespace aspect
+        AssertThrow(material_model->is_compressible() == true,
+                    ExcMessage("ASPECT detected an inconsistency in the provided input file. "
+                               "The mass conservation equation was selected to be compressible, "
+                               "but the provided material model reports to be incompressible. "
```
same here

bangerth commented on source/simulator/helper_functions.cc:
```diff
@@ -1320,6 +1323,106 @@ namespace aspect
+        AssertThrow(heating_model_manager.adiabatic_heating_enabled(),
+                    ExcMessage("ASPECT detected an inconsistency in the provided input file. "
+                               "The 'isothermal compression' formulation expects adiabatic heating to be enabled, "
+                               "but the heating model manager does not find a heating plugin "
```
does -> did

Alternatively, say that the 'adiabatic heating' plugin has not been selected in the input file, or similar.

bangerth commented on source/simulator/helper_functions.cc:
```diff
@@ -1320,6 +1323,106 @@ namespace aspect
+        AssertThrow(heating_model_manager.shear_heating_enabled(),
+                    ExcMessage("ASPECT detected an inconsistency in the provided input file. "
+                               "The 'isothermal compression' formulation expects shear heating to be enabled, "
+                               "but the heating model manager does not find a heating plugin "
```
same here

bangerth commented on source/simulator/helper_functions.cc:
```diff
@@ -1320,6 +1323,106 @@ namespace aspect
+      {
+        AssertThrow(!heating_model_manager.adiabatic_heating_enabled(),
+                    ExcMessage("ASPECT detected an inconsistency in the provided input file. "
+                               "The BA formulation expects adiabatic heating to be disabled, "
```
here and below, maybe expand the abbreviation or say "Boussinesq Approximation (BA)"? Same for the following blocks.

bangerth commented on source/simulator/parameters.cc:
```diff
@@ -302,6 +302,42 @@ namespace aspect
+                         Patterns::Selection ("isothermal compression|custom|ALA|BA"),
+                         "Select a formulation for the basic equations. Different "
+                         "published formulations are available in ASPECT (see list of "
+                         "possible parameters for available options). Two ASPECT specific "
```
see THE list of

But I'm actually not sure which list and where?

bangerth commented on source/simulator/parameters.cc:
```diff
@@ -302,6 +302,42 @@ namespace aspect
+                         "Select a formulation for the basic equations. Different "
+                         "published formulations are available in ASPECT (see list of "
+                         "possible parameters for available options). Two ASPECT specific "
+                         "options are: 'isothermal compression': ASPECT's original formulation, "
```
maybe make this into an enumerated list? This way you avoid the two colons in one line.

Use `\`` as an opening backtick for latex.

bangerth commented on source/simulator/parameters.cc:
```diff
@@ -302,6 +302,42 @@ namespace aspect
+                         "implemented for advanced users that want full control over the "
+                         "equations solved. It is possible to choose inconsistent formulations "
+                         "and no error checking is performed on the consistency of the resulting "
+                         "equations.");
```
So can I get the full equations (1)-(3) of the manual only with 'custom', which implies the warning?

gassmoeller replied:
No, the full equations (1)-(3) would be the 'isothermal compression' option. However so far our default was to either use the `isothermal compression` option or the incompressible version depending on the material model, and when writing this PR we decided that this behavior is somewhat dangerous (because we can not check if the user actually intends to run a compressible or incompressible model). Therefore we would like to require this setting in the parameter file in the future. For now though we kept the previous default behavior, which is equivalent to 'custom' with the associated warning.

bangerth commented on source/simulator/parameters.cc:
```diff
@@ -302,6 +302,42 @@ namespace aspect
+                         "Possible approximations for the density derivatives in the mass "
+                         "conservation equation. Note that this parameter is only evaluated "
+                         "if 'Formulation' is set to 'custom'. Other formulations overwrite "
+                         "this parameter.");
```
overwrite this parameter -> ignore the value of this parameter if specified in the input file.

(They do overwrite, but that's an implementation detail.)

bangerth commented on source/simulator/parameters.cc:
```diff
@@ -302,6 +302,42 @@ namespace aspect
+                         "Possible approximations for the density in the temperature equation. "
+                         "Possible approximations are: 'real density', 'reference profile'. "
+                         "Note that this parameter is only evaluated "
+                         "if 'Formulation' is set to 'custom'. Other formulations overwrite "
```
here and elsewhere, watch the opening single quotes and change to latex style.

bangerth commented on source/simulator/parameters.cc:
```diff
@@ -302,6 +302,42 @@ namespace aspect
+                         "conservation equation. Note that this parameter is only evaluated "
+                         "if 'Formulation' is set to 'custom'. Other formulations overwrite "
+                         "this parameter.");
+      prm.declare_entry ("Temperature equation", "real density",
```
I think for all of these parameters, we will eventually have to have a section in the manual that explains how they map to the various approximations in the new section you wrote as part of this PR. It's difficult for me to envision how these approximations are mapped to input parameters.

bangerth commented on source/simulator/parameters.cc:
```diff
@@ -949,6 +985,49 @@ namespace aspect
+
+          // Assert AdiabaticConditions = ConstantTemperature, density=1
+          // Assert shear/adiabatic heating plugins are off
+          // is_compressible= false
```
If you want to keep these, maybe prefix with a sentence that states that these checks happen in the `check_consistency...` function mentioned above.

Or add a sentence above where you introduce this function to say that each of the blocks below has a group of conditions that need to hold in each case, and that are actually checked in that function.

gassmoeller commented on include/aspect/assemblers.h:
```diff
@@ -0,0 +1,106 @@
+/*
+  Copyright (C) 2016 by the authors of the ASPECT code.
```
I think the idea here is to separate the concrete implementation of all the small assemblers (the function `local_assemble_...`) from the global tasks of assembling all equations. `assembly.h` and `assembly.cc` have become huge and completely confusing over time (more than 3000 lines in assembly.cc on master). 
The new structure in this PR is certainly not perfect either, but at least it separates the global stuff from the assembly of the individual terms.

This also relates to your comment further down about complexity of the assembler signals. I think (and we have discussed this with Timo and Juliane in Clemson) the signals have served their purpose, but in practice we do nothing with them that would not be possible with a vector of pointers to assembler objects. This could have exactly the same plugin structure we use everywhere else in the code, and we could remove the  the only signal we would keep would be the `set_assemblers` signal that allows plugins to modify the vector of pointers to `Assemblers::Interface` objects. This structure should be equally fast, and much easier to understand for users, because it follows the same design pattern as the rest of the code. This would obviously be a larger change that we should not mix with the current PR, but the separation into separate files might be a first step into that direction.
Would you agree with that general direction?

bangerth replied:
Let's just discuss the separation of files. What if you just separated out the *implementation* of assemblers into their own files -- say into `assembly_stokes.cc` and `assembly_advection.cc` but kept with a single header file?

gassmoeller replied:
I am ok with merging the files again for now (for this PR), but ultimately if we allow different implementations of these terms we should also split that visibly in different files. `assembly.cc` and `assembly.h` should only contain the not duplicative parts. I will merge assemblers.h and assembly.h and keep an old branch around for future plans.

jdannberg commented:
I can have a look at the benchmarks, I'll start with the King and TanGurnis. 
What do we want to do about Blankenbach? It would make sense to address the comments now anyway, so we might as well keep it as part of this pull request. But I can have a look and see if there's anything that's not really finished...

tjhei commented:
> What do we want to do about Blankenbach?

I ran two of the cases about a week ago (and added short notes in the .txt in that directory), but I am not confident that the whole benchmark is correct as is. I am hesitant to put ``cookbooks/blankenbach/*`` into master if we are not sure things are correct. Maybe we can find some time in the next two weeks to check and clean up.

gassmoeller commented on source/simulator/assemblers/stokes.cc:
```diff
@@ -0,0 +1,393 @@
+  template <int dim>
+  void
+  StokesAssembler<dim>::
+  local_assemble_stokes_compressible_diffusion (const double                                     /*pressure_scaling*/,
```
Yes it would be better to add some clarification to them. I was thinking about dropping the `local_assemble_stokes` part of the name, since all functions within `StokesAssembler` do that.

bangerth replied:
That works for me.

gassmoeller commented on source/simulator/assemblers/stokes.cc:
```diff
@@ -0,0 +1,393 @@
+            if (introspection.is_stokes_component(fe.system_to_component_index(i).first))
+              {
+                scratch.grads_phi_u[i_stokes] = scratch.finite_element_values[introspection.extractors.velocities].symmetric_gradient(i,q);
+                scratch.div_phi_u[i_stokes]   = scratch.finite_element_values[introspection.extractors.velocities].divergence (i, q);
```
I made some preliminary tests last week and it does not seem so. The speed of the old and new compressible scheme is nearly identical. A benefit of the new structure is that every assembler is only evaluating the test functions it needs. It is true that there is some redundancy, in particular for the `div_phi_u` terms, but that does not seem to slow things down significantly.

bangerth replied:
OK, if you know that it's not a performance problem, then that's fine with me.

It would just be a shame to undo all of the hard work you've done in finding performance bottlenecks!

gassmoeller commented on source/simulator/assemblers/stokes.cc:
```diff
@@ -0,0 +1,393 @@
+                                                         const Parameters<dim> &parameters) const
+  {
+    // assemble compressibility term of:
+    //  - div u - 1/rho * drho/dz g/||g||* u = 0
```
I do not think we have tested this extensively yet (it is also not included in any of the approximations), but it is indeed an interesting test to do. Should we leave it in the PR? Currently it is dead code.

bangerth replied:
If it's dead code, take it out and put it into a github issue for later inclusion. We know we have the structure right, so let's just make sure we get the things merged that we know actually work.

gassmoeller replied:
I just saw that there are two tests using this formulation. @tjhei: What is your take on this option? Is it tested well enough to be kept in the code for merge, or would it need further testing?

tjhei replied:
It works correctly. I have two tests, one that uses the explicit form and one that uses this implicit form. The latter convergence in one Newton iteration, so I am pretty sure it works. As to the linear solver: yes, it might be worse and it needs more testing. If you look at  ``tangurnis_tala[_implicit]/screen_output``, they both need 30+2 iterations.

gassmoeller commented on source/simulator/assembly.cc:
```diff
@@ -2780,11 +1123,13 @@ namespace aspect
-//TODO: the following doesn't currently compile because the get_quadrature() call returns
+
+              // TODO: formulations?
+
```
I Have update the TODO to the following:      

              // TODO: Currently we do not supply reference density values to Stokes face assemblers.
              // This seems acceptable for now, since the only face assemblers are the ones for the melt
              // assembly where one would not want to use the reference density anyway. In case the reference
              // density is ever needed those assemblers can also query the adiabatic conditions for the
              // reference density.

All in all this TODO shows a design decision we might need to think about at some point. Is it the responsibility of `Simulator::local_assemble_stokes_system` to provide every necessary information to the assemblers inside of the `scratch` object, or should it only provide the basic information, and in any case an assembler needs more information it is responsible for getting this information? For now  `Simulator::local_assemble_stokes_system` provides or manipulates the information it gives to the assemblers in order to ensure the approximations (and following this design we would need to introduce `scratch.mass_density_face` variables and fill them for the face assemblers).

gassmoeller commented on source/simulator/entropy_viscosity.cc:
```diff
@@ -0,0 +1,439 @@
+
+        if (advection_field.is_temperature())
+          {
+            max_density = std::max       (scratch.material_model_outputs.densities[q],              max_density);
```
We already do. `Simulator<dim>::get_artificial_viscosity` replaces the real density by the reference density if we use the reference density profile (entropy_viscosity.cc line 383)

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1034,6 +1034,15 @@ \subsection{Static or dynamic pressure?}
+which makes it possible to formulate the Stokes system in terms of the dynamic instead of
+the full pressure. In order to do this, one would have to use a material model 
+(see Section~\ref{sec:material-models}) in which the density is in fact a density variation, 
+and the the pressure solution variable would only be the dynamic pressure.}
```
and THEN the?

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1339,21 +1354,234 @@ \subsection{Numerical methods}
+where $\bar\rho=\rho(\bar p,\bar T)$ is a \textit{reference density} that
+depends on depth and represents a typical change of material parameters and solution 
+variables with depth. $\bar T(z)$ is chosen as an adiabatic profile accounting for the
+fact that the temperature increases as the pressure increases.
```
would it be worthwhile pointing out that `\bar \rho` can be chosen self-consistently as an adiabatic profile as well?

bangerth replied:
Or do you want to wait until you talk about the individual approximations with this?

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1339,21 +1354,234 @@ \subsection{Numerical methods}
+  \quad
+  \textrm{in $\Omega$},
+\end{multline}
+where the right-hand side includes radiogenic heat production, shear heating and adiabatic heating (in that order).
```
how does this equation differ from the original one at all? (Other than the omission of entropic heating?)

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1339,21 +1354,234 @@ \subsection{Numerical methods}
+perturbations $p'$, $T'$ are much smaller than typical values of the reference
+quantities $\bar p$, $\bar T$. 
+The terms influenced by these approximations are $\nabla \cdot (\rho u) =0$ in the 
+continuity equation, and all occurrences of $\rho(p,T)$ in the temperature equation, 
```
but it also affects the occurrence of `\rho` in the force balance equation, right?

tjhei replied:
No, we always use the real rho coming from the material model in the buoyancy term.

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1339,21 +1354,234 @@ \subsection{Numerical methods}
+\end{multline}
+
+\note{Our energy equation is formulated in terms of $T$, while in the literature, people 
+have chosen to formulate the equation in terms of $T'$, which yields additional terms 
```
"people" is a bit informal. How about "...while in the literature, the equation has sometimes been formulated in terms of $T'$, then yielding additional terms..."

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1517,55 +1669,49 @@ \subsubsection{Almost linear models}
+arrives at equations in which the only nonlinear term is the advection term,
+$\mathbf u \cdot \nabla T$ in the temperature equation
 \eqref{eq:temperature-boussinesq-linear}. This facilitates the use of a
 particular class of time stepping scheme in which one does not solve the whole
```
not your fault, but while you're there: -> time stepping SCHEMES (plural)

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1419,94 +1645,20 @@ \subsubsection{The Boussinesq approximation: Incompressibility}
-
-
-\subsubsection{Almost linear models}
+\paragraph{On almost linear models.}
```
It seems like this paragraph has mostly become obsolete. Or do you think it's worth keeping?

jdannberg replied:
We just discussed that a few days ago, and you suggested (and we agreed) to leave it in there because it is worth mentioning that the time stepping scheme has changed compared to older Aspect versions and in particular to the first paper. I removed everything else from that paragraph.

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1517,55 +1669,49 @@ \subsubsection{Almost linear models}
+
+\note{\aspect{} does not solve the equations in the way described in this paragraph, 
+however, a particular operator splitting method was used in
+earlier \aspect{} versions. It first solves the Stokes equations and then
```
That's a very long sentence. Maybe split in two?

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1577,7 +1723,7 @@ \subsubsection{Compressible models}
+Earth's mantle, the density increases by at least 50\% between Earth's crust and
 the core-mantle boundary due to larger pressure there. Secondly, we assume
 that the pressure is dominated by the static pressure, which can be written as
 $\nabla p \approx \nabla p_s \approx \rho \textbf{g}$. This is essentially
```
can be written -> implies that

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1595,14 +1741,198 @@ \subsubsection{Compressible models}
 where $\frac{1}{\rho} \frac{\partial \rho}{\partial p}$ is often referred to
 as the compressibility.
```
I don't think I understand where that last sentence comes in. It's also a bit awkward ("essentially").

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1595,14 +1741,198 @@ \subsubsection{Compressible models}
 
-In the implementation used in \aspect{}, this equation replaces
+For this approximation, this equation replaces
 \eqref{eq:stokes-2}. It has the advantage that it retains the symmetry of the
```
There's no *equation* above, just an *identity*. The equation would require that the term be zero.

But it is also not symmetric to the other term, only if you bring the second term to the rhs.

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1595,14 +1741,198 @@ \subsubsection{Compressible models}
+
+
+\newpage
+\subsection{Choosing a formulation in \aspect{}}
```
no need for the newpage -- let latex deal with these questions

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1595,14 +1741,198 @@ \subsubsection{Compressible models}
+\newpage
+\subsection{Choosing a formulation in \aspect{}}
+
+After discussing different reasonable approximations for modelling compressible or 
```
-> modeling (one 'l' in American English)

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1595,14 +1741,198 @@ \subsubsection{Compressible models}
+This includes 
+\begin{enumerate}
+\item Choosing an approximation for the mass conservation equation, 
+\item Choosing an approximation for the density in the energy balance, and deciding which heating terms should be included, 
```
end each item with a semicolon

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1595,14 +1741,198 @@ \subsubsection{Compressible models}
+\item Choosing an approximation for the mass conservation equation, 
+\item Choosing an approximation for the density in the energy balance, and deciding which heating terms should be included, 
+\item Formulating the buoyancy term in the material model to be used on the right-hand side of the momentum equation, 
+\item Prescribing a suitable reference state for the temperature, pressure, and density, the adiabatic profile.
```
do you mean "...and density; i.e., the adiabatic profile"?

Maybe also add "..., if necessary for the approximations chosen in the first three steps".

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1595,14 +1741,198 @@ \subsubsection{Compressible models}
+
+\item
+``incompressible'':
+\[
```
latex guides strongly recommend not to use `\[` or `$$` but use `\begin{equation*}` instead because that allows latex to use the style chosen for a document, rather than some low-level tex stuff. Same below.

bangerth replied:
Well, actually, I take that back -- http://tex.stackexchange.com/questions/40492/what-are-the-differences-between-align-equation-and-displaymath . It's ok to use `\[`.

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1595,14 +1741,198 @@ \subsubsection{Compressible models}
+\item
+``isothermal compression'':
+\[
+ \nabla \cdot \textbf{u} = -\frac{1}{\rho} \frac{\partial \rho}{\partial p} \rho \textbf{g} \cdot \textbf{u},
```
this corresponds to the "explicit" option, right? Or do you not want to dwell on this difference? (Is there even a way to select whether you want the explicit or implicit option?)

jdannberg replied:
There is only the explicit version for now (although our results for ALA suggest we might want to think about adding the implicit one some time in the future).

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1595,14 +1741,198 @@ \subsubsection{Compressible models}
+\subsubsection{Mass conservation approximation}
+
+First, we have to choose how to approximate the conservation of mass: $\nabla \cdot (\rho \mathbf u) = 0$ (see Equation~\eqref{eq:stokes-2}). 
+We provide the following options:
```
This is still in the section on equations and models. What I think would be useful to add here is a reference to 
* the fact that these choices all correspond to parameters that can be selected in the input file
* the section that explains the basics of running aspect
* the section in the appendix that lists the relevant parameters

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1595,14 +1741,198 @@ \subsubsection{Compressible models}
+ \nabla \cdot \textbf{u} = -\frac{1}{\rho} \frac{\partial \rho}{\partial p} \rho \textbf{g} \cdot \textbf{u},
+\]
+where the compressibility $\kappa = \frac{1}{\rho} \frac{\partial \rho}{\partial p}$ 
+is defined in the material model. 
```
This `\kappa` doesn't actually appear in the formula. I think it would be useful to put it there.

jdannberg replied:
I think the way we write it like now makes it easier to see the differencesame between the formulations, but we'll rewrite the description.

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1595,14 +1741,198 @@ \subsubsection{Compressible models}
+\[
+ \nabla \cdot \textbf{u} = -\frac{1}{\bar{\rho}} \frac{\partial \bar{\rho}}{\partial z} \frac{\textbf{g}}{\|\textbf{g}\|} \cdot \textbf{u},
+\]
+where reference profiles for the density $\bar{\rho}$ and the density gradient $\frac{\partial \bar{\rho}}{\partial z}$
```
where THE reference

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1595,14 +1741,198 @@ \subsubsection{Compressible models}
+\]
+where reference profiles for the density $\bar{\rho}$ and the density gradient $\frac{\partial \bar{\rho}}{\partial z}$
+provided by \text{AdiabaticConditions}
+are used. Note that the
```
I don't understand what `\text{AdiabaticConditions}` is supposed to reference. (Also, `\texttt`?) Should this be a reference to a section? A parameter?

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1595,14 +1741,198 @@ \subsubsection{Compressible models}
+where reference profiles for the density $\bar{\rho}$ and the density gradient $\frac{\partial \bar{\rho}}{\partial z}$
+provided by \text{AdiabaticConditions}
+are used. Note that the
+gravity is assumed to point downwards in depth direction
```
period at the end of the sentence missing.

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1595,14 +1741,198 @@ \subsubsection{Compressible models}
+which uses the same approximation for the density as ``reference density profile'', 
+but implements this term on the left-hand side instead of the right-hand side of the mass
+conservation equation. This effectively uses the current velocity $\textbf{u}$ instead
+of an extrapolated velocity from the last timesteps.
```
this suggests that we should have made clearer in the previous equations that things on the right hand side are extrapolated, e.g., by using `^\ast` on the velocity and pressure.

jdannberg replied:
I think the way it is now it's easier to understand. This is also something that should be consistent throughout the manual, so there would be a lot of other places where we would have to think about changing it.

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1595,14 +1741,198 @@ \subsubsection{Compressible models}
+
+ \item 
+``ask material model'', which uses ``isothermal compression'' if the material model is compressible 
+(as reported by the \texttt{is\_compressible()} function) and ``incompressible'' otherwise.
```
This references a particular implementation detail that is not yet available at this point in the manual. I would just remove the parenthetical remark.

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1595,14 +1741,198 @@ \subsubsection{Compressible models}
+\[
+  \nabla \cdot \tau =  \nabla \cdot \left[
+ 2\eta \varepsilon(\mathbf u) \right].
+\]
```
Do you want to say that for certain choices of the mass conservation parameters discussed above, this simplification is automatically used in the force balance equation?

bangerth replied:
Right now, the `\paragraph` ends a bit abruptly without a proper conclusion :-)

jdannberg replied:
Yes. If the mass conservation equation is incompressible, \nabla \cdot \mathbf u = 0 anyway, and we don't need to assemble the term in other places (because it is zero). 
Also, I'll make it a note instead of a paragraph, and then it doesn't need a conclusion.

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1595,14 +1741,198 @@ \subsubsection{Compressible models}
+
+\subsubsection{Temperature equation approximation}
+
+The density occurs multiple times in the temperature equation. Depending on the selected approximation it is computed in two different ways:
```
same here -- it would be worthwhile stating *how* one can select an approximation

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1595,14 +1741,198 @@ \subsubsection{Compressible models}
+
+\begin{itemize}
+ \item 
+``real density'': Use the full density as computed for the buoyancy term of the momentum equation (i.e. the return value of the material model for ``density``),
```
Maybe state this as "Use the full density $\rho(p,T)$ that equals the one also used in the buoyancy term of the force balance equation; this is also the value that is computed by the material models when asked for the density."

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1595,14 +1741,198 @@ \subsubsection{Compressible models}
+\subsubsection{Approximation of the buoyancy term}
+The buoyancy term (right-hand side of the momentum equation) always uses the 
+density that is provided by the material model (see Section~\ref{sec:material-models}). 
+There, one can decide to use a density that for example depends on temperature 
```
"There" -- where? You mean in the material model?

It's also not quite clear what "one can decide" means. I suppose you mean that one can decide this when implementing a new material model?

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1595,14 +1741,198 @@ \subsubsection{Compressible models}
+
+The reference temperature profile $\bar{T}$, reference density profile $\bar{\rho}$
+and the reference pressure $\bar{p}$ are computed in the adiabatic conditions model 
+(provided by the class \textit{AdiabaticConditionds}, see Section~\ref{sec:adiabatic}). 
```
texttt

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1595,14 +1741,198 @@ \subsubsection{Compressible models}
+and the reference pressure $\bar{p}$ are computed in the adiabatic conditions model 
+(provided by the class \textit{AdiabaticConditionds}, see Section~\ref{sec:adiabatic}). 
+By default, these fields satisfy adiabatic conditions (if adiabatic heating is included 
+in the model):
```
how does one decide whether adiabatic heating is included?

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1595,14 +1741,198 @@ \subsubsection{Compressible models}
+  &=
+  \bar\rho g_z,
+\\
+  \frac{\textrm{d} \bar{p}}{\textrm{d}z} &= \bar\rho g_z,
```
You have the same equation for `\bar p` twice here. Did you mean to have one of them be for the density?

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1595,14 +1741,198 @@ \subsubsection{Compressible models}
+\begin{align}
+  \rho C_p \frac{\textrm{d} \bar{T}(z)}{\textrm{d}z} 
+  &=
+  \frac{\partial\bar\rho}{\partial \bar{T}} \bar{T}(z) g_z,
```
should be `\frac{\partial\bar\rho}{\partial T}` (not `\bar T` in the denominator).

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1595,14 +1741,198 @@ \subsubsection{Compressible models}
+magnitude of the entire gravity vector.
+If there is no adiabatic heating in the model, $\bar{T}$ is constant 
+by default and set to the adiabatic surface temperature.
+The density gradient is then computed by a simple finite difference approximation 
```
add a paragraph break before this sentence, or some other way, to make clear that it doesn't refer to the sentence before, but to the formulas above.

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1595,14 +1741,198 @@ \subsubsection{Compressible models}
+component of the gravity vector field, but in practice we take the
+magnitude of the entire gravity vector.
+If there is no adiabatic heating in the model, $\bar{T}$ is constant 
+by default and set to the adiabatic surface temperature.
```
shouldn't alpha appear in one of these equations above?

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1595,14 +1741,198 @@ \subsubsection{Compressible models}
+However, users can also supply their own adiabatic conditions models or 
+define an arbitrary profile using the ``function'' plugin, 
+which allows the user to define arbitrary functions for 
+$\bar{T}(z)$, $\bar{p}(z)$ and $\bar{\rho}(z)$.
```
, see Section~\ref{...}

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1595,14 +1741,198 @@ \subsubsection{Compressible models}
+$\bar{T}(z)$, $\bar{p}(z)$ and $\bar{\rho}(z)$.
+
+\subsubsection{Combined Formulations}
+\label{sec:combind_formulations}
```
typo in the label

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1595,14 +1741,198 @@ \subsubsection{Compressible models}
+which allows the user to define arbitrary functions for 
+$\bar{T}(z)$, $\bar{p}(z)$ and $\bar{\rho}(z)$.
+
+\subsubsection{Combined Formulations}
```
lower case "formulations" to be consistent with the style used elsewhere

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1595,14 +1741,198 @@ \subsubsection{Compressible models}
+\subsubsection{Combined Formulations}
+\label{sec:combind_formulations}
+Not all combinations of the different approximations discussed above 
+are physically reasonable, and to help users to choose between these options, 
```
-> help users choose

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1595,14 +1741,198 @@ \subsubsection{Compressible models}
+Not all combinations of the different approximations discussed above 
+are physically reasonable, and to help users to choose between these options, 
+we provide a number of combined ``Formulations'' that are equivalent to the 
+approximate equations discussed above (Section~\ref{sec:approximate-equations}):
```
as above, it would be useful to link to the place where one can select these

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1595,14 +1741,198 @@ \subsubsection{Compressible models}
+The default setting for the adiabatic conditions is an adiabatic temperature profile, and hydrostatic
+pressure and density profiles. This option should be chosen together with a material model 
+that defines a density that depends on temperature and pressure (and potentially depth), 
+which would be equivalent to the Anelastic liquid approximation (Section~\ref{sec:ala}), 
```
uniformly capitalize ALA (or not capitalize at all)

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1595,14 +1741,198 @@ \subsubsection{Compressible models}
+that defines a density that depends on temperature and pressure (and potentially depth), 
+which would be equivalent to the Anelastic liquid approximation (Section~\ref{sec:ala}), 
+or with a material model that defines a density that depends on temperature and depth 
+(and not on the pressure), which would be equivalent to the Truncated anelastic liquid approximation 
```
same here

bangerth commented:
We really ought to write a cookbook that shows how to use the various ways to select formulations. But that can wait for a separate PR.

bangerth commented on source/postprocess/heating_statistics.cc:
```diff
@@ -114,6 +114,11 @@ namespace aspect
 
             this->get_material_model().evaluate(in, out);
 
+            if (this->get_parameters().formulation_temperature_equation
```
can you add a one-line comment that says that you're explicitly overwriting the density from the material model here if the formulation requires this?

tjhei replied:
Sure. I also added an else block with the other option (with an empty block) and an else ``ExcNotImplemented`` in case we ever extend the number of enum options.

bangerth commented on source/postprocess/visualization/heating.cc:
```diff
@@ -144,6 +144,11 @@ namespace aspect
 
         this->get_material_model().evaluate(in, out);
 
+        if (this->get_parameters().formulation_temperature_equation
```
same here

bangerth commented on source/simulator/entropy_viscosity.cc:
```diff
@@ -0,0 +1,438 @@
+
+        material_model->evaluate(scratch.material_model_inputs,scratch.material_model_outputs);
+
+        if (parameters.formulation_temperature_equation ==
```
and here

bangerth commented on source/simulator/parameters.cc:
```diff
@@ -302,6 +302,42 @@ namespace aspect
+                         "Select a formulation for the basic equations. Different "
+                         "published formulations are available in ASPECT (see the list of "
+                         "possible values for this parameter in the manual for available options). "
+                         "Two ASPECT specific options are 1. `isothermal compression': ASPECT's original "
```
Can you make this into an `enumerate` latex environment? This would make it much easier to read.

bangerth commented on source/simulator/parameters.cc:
```diff
@@ -302,6 +302,42 @@ namespace aspect
+                         "possible values for this parameter in the manual for available options). "
+                         "Two ASPECT specific options are 1. `isothermal compression': ASPECT's original "
+                         "formulation, using the explicit compressible mass equation, "
+                         "and the full density for the temperature equation. 2. 'custom': "
```
backtick before "custom"

bangerth commented on source/simulator/parameters.cc:
```diff
@@ -302,6 +302,42 @@ namespace aspect
+                         "formulation, using the explicit compressible mass equation, "
+                         "and the full density for the temperature equation. 2. 'custom': "
+                         "A custom selection of `Mass conservation' and "
+                         "`Temperature equation'. Warning: The `custom' option is "
```
Put the warning into a `\note{...}` environment to make it stick out.

bangerth commented on source/simulator/parameters.cc:
```diff
@@ -302,6 +302,42 @@ namespace aspect
+      prm.declare_entry ("Temperature equation", "real density",
+                         Patterns::Selection ("real density|reference density profile"),
+                         "Possible approximations for the density in the temperature equation. "
+                         "Possible approximations are: `real density', `reference profile'. "
```
-> reference DENSITY profile

bangerth commented on tests/tangurnis.cc:
```diff
@@ -99,7 +100,7 @@ namespace aspect
          * does not necessarily imply that the density is constant; rather, it
          * may still depend on temperature or pressure. In the current
-         * context, compressibility means whether we should solve the continuity
+         * context, compressibility means whether we should solve the contuity
```
wrong change here

bangerth commented on tests/tangurnis.cc:
```diff
@@ -462,63 +463,28 @@ namespace aspect
+                                   const Point<dim> &position) const
+      {
+        double wavenumber=1;
+        return sin(numbers::PI*position(dim-1))*cos(numbers::PI*wavenumber*position(0));
```
can you use `std::sin` and `std::cos` here?

tjhei commented:
YAY!

gassmoeller commented:
@GEuen: The new features are merged into the latest master. Feel free to go ahead with your testing.
